### ISSUE TITLE
Ruby ternary operator and regex highlighting 

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#69]
Created for https://sourceforge.net/p/scintilla/bugs/1927/ and https://sourceforge.net/p/scintilla/bugs/2126/

Here need update for `?` (question mark single character string) and `:` (symbol).
https://ruby-doc.org/core-3.1.1/doc/syntax/literals_rdoc.html#label-Strings
https://ruby-doc.org/core-3.1.1/doc/syntax/literals_rdoc.html#label-Symbol+Literals

**Comment by zufuliu** [comment]
[ruby-ternary-0327.zip](https://github.com/ScintillaOrg/lexilla/files/8356800/ruby-ternary-0327.zip)

Contains following changes:
1. for `:`, treat it as operator instead of symbol if followed by space
2. for `?`
    1. if followed by an identifier with at least two code points, then treat it as operator instead of number. e.g. `?ab` is colored as operator `?` with identifier `ab`.
    2. correctly handle multi-byte character after it with `GetCharacterAndWidth()`, which fixed highlighting bug for `?あ` from linked doc.

As show in the test file, this is not full fixed, I think maybe further complex work is needed to handle ternary operator not followed by space (not a good code style though).
```ruby
# TODO: space after '?' and ':' is not needed
puts true ?ab : cd
puts true ? /d+/ : /w+/
puts false ?ab : cd
puts false ? /d+/ : /w+/
```

**Comment by nyamatongwe** [comment]
One of the hassles with `noexcept` is that it's viral.
~~~
lexilla\lexers\LexRuby.cxx(69): warning C26447: The function is declared 'noexcept' but calls function 'isSafeAlnum()' which may throw exceptions (f.6).
lexilla\lexers\LexRuby.cxx(70): warning C26447: The function is declared 'noexcept' but calls function 'isSafeWordcharOrHigh()' which may throw exceptions (f.6).
~~~

The original LexRuby.cxx uses space indentation so its best to maintain that.

Can't see how it is supposed to disambiguate `?` and `:` when there are no spaces.

**Comment by zufuliu** [comment]
[ruby-ternary-0329-3.zip](https://github.com/ScintillaOrg/lexilla/files/8367008/ruby-ternary-0329-3.zip)

Removed `noexcept` and converted indentation to spaces. changed regex to `/\d+/` and  `/\w+/`.

I also not find how to disambiguate them, but `?/\` and `:/\` are invalid when `\` not followed by new line, maybe can be special handled.

**Comment by nyamatongwe** [comment]
For line 5, `puts ?\\` the first byte of the following newline is still in `SCE_RB_NUMBER`. I would expect the 'number' to include just the two `\`. 

This hasn't changed but the example will cause trouble with TestLexers due to the bytes of `\r\n` having different styles. Making the example line `puts ?\\#` would avoid warnings from TestLexers until the underlying problem is fixed.

**Comment by zufuliu** [comment]
[ruby-ternary-0401.zip](https://github.com/ScintillaOrg/lexilla/files/8395356/ruby-ternary-0401.zip)

Fixed `?\\`.

**Comment by nyamatongwe** [comment]
Pushed in 2 commits listed above.

**Comment by nyamatongwe** [comment]
Included in 5.1.7 release.
